feat: Handle deposits that do not map to a PoolRebalanceRoute mapping (#619)
UMAprotocol/UMIPs: UMIPs/umip-179.md

3. The corresponding `Deposit` event occurred within or before the `Bundle Block Range` on the origin chain SpokePool.

#### Note

> - If the `Deposit` event specifies `outputToken` `bytes32(0)` (i.e. the Zero Address), the equivalent SpokePool token on the destination chain shall be substituted in. For the purpose of determining `RelayData` equivalency, the updated/substituted `outputToken` shall be used in place of the Zero Address. If there is no equivalent SpokePool token on the destination chain at the `Deposit` `quoteTimestamp`, then this deposit cannot be filled successfully. This means that any fills or slow fill requests sent for this deposit cannot be validated and will not be refunded.

Author: I think the UMIP should block these fills from being repaid, and then the dataworker should add its own protection as we've done [here](https://github.com/across-protocol/sdk/pull/970#discussion_r2045511985)

> - If the `Deposit` event specifies `outputToken` `bytes32(0)` (i.e. the Zero Address), the equivalent SpokePool token on the destination chain shall be substituted in. For the purpose of determining `RelayData` equivalency, the updated/substituted `outputToken` shall be used in place of the Zero Address. If there is no equivalent SpokePool token on the destination chain at the `Deposit` `quoteTimestamp`, then this deposit cannot be filled successfully. This means that any fills or slow fill requests sent for this deposit cannot be validated and will not be refunded.

Reviewer: Agreed. Since the DW blocks this, I think I would tweak the UMIP language to `this deposit will not be repaid regardless of whether it is fillable on the destination.`

It's a nit, but since the DW invalidates it regardless, we should specify that here, too.

> The applied `repaymentChainId` shall be determined as follows:

Author: I like deposit quote timestamp

> The applied `repaymentChainId` shall be determined as follows:

Author: 55826194bf0c6e68776cba7c64b3183d751bbc5a

> The applied `repaymentChainId` shall be determined as follows:

Author: we would use the mainnet bundle end block. This is a good point, we actually can't really use deposit quote timestamp. Otherwise there's a possibility that a deposit gets marked as "can be repaid" but the pool rebalance route was deleted by the end of the bundle and we can't actually pay them back

- When the `originChainId` is a `Lite chain` as at the `Deposit` `quoteTimestamp`: `originChainId`, ELSE
- When no `PoolRebalanceRoute` exists for the `inputToken` on the `Fill` `originChainId` as at the `Deposit` `quoteTimestamp`: `originChainId`, ELSE

> - When no `PoolRebalanceRoute` exists for the `repaymentToken` on the `Fill` `repaymentChainId` as at the `Deposit` `quoteTimestamp`: `originChainId`, ELSE

Author: good point, clarified it a bit here: ad0ca526af6ac48b90acce1318a6d03bd0575902

- The `repaymentChainId` as specified in the `Fill`.

The applied `repayment address` shall be determined as follows:
- When the `Fill` `relayer` address is valid for the applied `repaymentChainId`: `relayer`, ELSE

> - The `Fill` `msg.sender` address. In this case, change the applied `repaymentChainId` to the `destinationChainId` if a `PoolRebalanceRoute` exists for both the `outputToken` on the `destinationChainId` and the `inputToken` on the `originChainId`.

Reviewer 2: I'm a bit uncertain on this change. `repaymentChainId` should already be set according to `PoolRebalanceRoute` considerations, and `repayment address` is intended to be downstream of this. It seems awkward to me to circle back and change the `repaymentChainId` here.

`msg.sender` also doesn't work for `destinationChainId` where origin & destination chain families are not compatible (i.e. evm/svm).

Author: The situation I'm trying to capture with this change is:
- repaymentChainId is a valid chain given the Deposit.inputToken and Deposit.originChain
- However, `repaymentAddress` is an invalid EVM address (e.g. its a 32 byte SVM address)
- (I believe) we should then attempt to change `repaymentAddress` to the `msg.sender` and therefore also change the `repaymentChain` to the `destinationChain`

> msg.sender also doesn't work for destinationChainId where origin & destination chain families are not compatible (i.e. evm/svm).

Agreed, and so this is why I added the line: "if a `PoolRebalanceRoute` exists for both the `outputToken` on the `destinationChainId` and the `inputToken` on the `originChainId`"

WDYT?

> - The `Fill` `msg.sender` address. In this case, change the applied `repaymentChainId` to the `destinationChainId` if a `PoolRebalanceRoute` exists for both the `outputToken` on the `destinationChainId` and the `inputToken` on the `originChainId` as at the `Deposit` `quoteTimestamp`.

Reviewer 3: This means that we can reach a path where we refund `msg.sender` on `originChainId`, which may not exist if `msg.sender` is a contract on destination. Is this fine? Should we add a third `ELSE` which marks the fill as unrepayable?

Reviewer 3: Never mind. I think it's probably just better to refund here.

- See examples [here](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/742e85be7c08dff21410ba4aa9c60f6a033befb8/test/utils/cryptography/MerkleProof.test.js) for how to construct these types of trees.

### Constructing the Relayer Refund Root
At least one Relayer Refund Leaf shall be produced for each unique combination of SpokePool and `repaymentToken` for any of the following conditions:
- Valid `Fills`, OR
- Expired `Deposits`, OR
- Unfillable `Deposits`, OR
- A negative running balance net send amount.

> Where `repayment` token is determined as follows for

Reviewer 2: ```suggestion
Where `repaymentToken` is determined as follows for
```

> - `Fills`: the [equivalent](#resolving-spokepool-tokens-to-their-hubpool-equivalent) L2 token address for the `repaymentChainId`, as computed [here](#computing-relayer-repayments), of the `Fill.
> - `Deposits`: the `inputToken`
> - negative running balance net send amounts: the L2 token address for the corresponding `l1Token` considered in Pool Rebalance Root production

Reviewer 2: Not super keen on using `L2` terminology here since we're also supporting alternative L1s.

```suggestion
- `Fills`: the [equivalent](#resolving-spokepool-tokens-to-their-hubpool-equivalent) token address for the `repaymentChainId`, as computed [here](#computing-relayer-repayments), of the `Fill.
- `Deposits`: the `inputToken`
- negative running balance net send amounts: the token address for the corresponding `l1Token` considered in Pool Rebalance Root production
```

Each Relayer Refund Leaf shall be constructed as follows:
- `amountToReturn` shall be set to `max(-netSendAmount, 0)`.

> - `l2TokenAddress` shall be set to the `repayment` token as computed previously

Reviewer 2: ```suggestion
- `l2TokenAddress` shall be set to the `repaymentToken` as computed previously
```
